paddle/operators/deconv2d_op.h

```diff
@@ -0,0 +1,251 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include "glog/logging.h"
+#include "paddle/framework/eigen.h"
+#include "paddle/framework/op_registry.h"
+#include "paddle/operators/math/im2col.h"
+#include "paddle/operators/math/math_function.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+using DDim = framework::DDim;
+
+// Define Op classes in .h file so that other deconv
+// operator implementations can reuse the code.
+class Deconv2DOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  Deconv2DOpMaker(framework::OpProto* proto,
+                  framework::OpAttrChecker* op_checker);
+};
+
+class Deconv2DOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(framework::InferShapeContext* ctx) const override;
+};
+
+class Deconv2DOpGrad : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(framework::InferShapeContext* ctx) const override;
+};
+
+template <typename Place, typename T>
+class GemmDeconv2DKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    const Tensor* input = context.Input<Tensor>("Input");
+    // filter will be reshaped, so we do not use constant pointer here
+    Tensor filter = *context.Input<Tensor>("Filter");
+
+    Tensor* output = context.Output<Tensor>("Output");
+
+    std::vector<int> strides = context.Attr<std::vector<int>>("strides");
+
+    // no paddings and groups allowed in deconv
+
+    int N = input->dims()[0];
+    int M = input->dims()[1];
+    int H = input->dims()[2];
+    int W = input->dims()[3];
+
+    int K_H = filter.dims()[2];
+    int K_W = filter.dims()[3];
+
+    int C = output->dims()[1];  // output channels
+    int O_H = output->dims()[2];
+    int O_W = output->dims()[3];
+
+    paddle::operators::math::Col2ImFunctor<
+        paddle::operators::math::ColFormat::kCFO, Place, T>
+        col2im;
+
+    // use col_shape in the im2col and col2im calculation
+    DDim col_shape = {C, K_H, K_W, H, W};
+
+    // use col_matrix_shape in the gemm calculation
+    DDim col_matrix_shape = {C * K_H * K_W, H * W};
+
+    Tensor col;
+    col.mutable_data<T>(col_shape, context.GetPlace());
+    // col_matrix shares the same piece of data with col,
+    // but will be reshaped into a two-dimensional matrix shape
+    // to call the matrix multiplication interface.
+    Tensor col_matrix = col;
+    col_matrix.Resize(col_matrix_shape);
```

Comment:
That copy assign works as intended, but i looks a little unnatural to me at first glance, since for e.g. std::vector<T>, copy assign copies the data. However, copy assignment does share data in this case because the data is stored inside a  std::shared_ptr<Placeholder> inside the Tensor class. Nevertheless, I would suggest the more explicit:
```
Tensor col_matrix;
col_matrix.ShareDataWith(col);
```
(I realize this is carried over from conv2d_op.h - maybe you could change it there, too?)

Reply:
Great thanks!

Reply:
Done.